### DISCUSSION TITLE
Rustic support in Backrest

### DISCUSSION THREAD

**Discussion by garethgeorge** [rustic-rs/rustic#1160] [General]
Hey, [backrest](https://github.com/garethgeorge/backrest) developer here! Backrest is a multiplatform WebUI for restic. Discovered the rustic project some months ago and as it matures I'm interested to learn more about it / whether it's a good option to support in the WebUI. 

Reaching out to learn more about the state of the project and whether rustic is a backend that I might support in the backrest project. To support a new backend it generally needs to support the operations in [restic.go](https://github.com/garethgeorge/backrest/blob/main/pkg/restic/restic.go) in the backrest project.

A quick search of the repo shows that there is some amount of support for JSON output in the code base (which is awesome to see!). Backrest works by forking restic and parsing it's JSON mode output for many commands. I'm interested to hear how close to restic rustic is and what major differences there might be that I'll run into. I'm thinking compatibility may be very easy if there's generally a 1:1 correspondence between the commands and generally the data available in the JSON output from commands (even if the fields are difference this can be addressed by converters). 

I notice that rustic also offers a library -- linking against rustic might be an option here too! Questions I have here would mostly be around how I can best capture and surface any rustic logs etc if calling core.

**Comment by aawsome** [comment]
Hi  @garethgeorge!
Thanks for you interest in rustic! In principle, rustic is not 100% CLI compatible to restic, but very similar. Most commands work like restic or at least very similar. 
So, if you want to use rustic by calling the CLI, I would say just try it out - it will work pretty much like restic.

There are a few things to note however, which are mainly due to the differences of restic and rustic, see https://rustic.cli.rs/docs/comparison-restic.html:
- Configuring remote repositories is quite different as rustic uses `opendal` for this. On the other side, more backends are supported directly compared to restic
- If you want to use hot/cold repositories, this also needs extra configuration and commands may work differently.
- Some commands do not exist while others have additional functionality
- Note that rustic is lock-free

A better option would be to use the `rustic_core` crate where you have much finer control over the repository commands and direct output, see https://docs.rs/rustic_core/latest/rustic_core/. But I fear the main point here would be to make it accessible to a Golang project. Besides this, progress information is basically available by setting a `ProgressBar` to the `Repository` which is a callback-function. logs are created using the `log` crate, i.e. you can define your own log handler (again basically a call-back) and do whatever you want.

**Reply by garethgeorge** [reply]
Hi, realized I forgot to followup on this thread :) thanks for the detailed write up! Looking over the rustic API and it's capabilities looks quite exciting -- and I expect if integrated with Backrest it should offer _significantly_ better performance than restic as, with control of the life cycle, I can keep rustic alive as a daemon across multiple operations (I envisage neat features like real time file browsing, etc).

[rustic_core](https://docs.rs/rustic_core/latest/rustic_core/) looks great for my needs and is probably the way to go for Backrest -- seems like it'll give a lot of control BUT it does make backrest responsible (at a very low level) for constructing configuration and sequencing operations. The latter bit -- sequencing operations is likely fine, this is most of what Backrest does :) and if you agree, it probably makes sense to take an opinionated approach that's similar to what Backrest does today for restic.

For IPC, Backrest uses gRPC for the WebUI to communicate to it's backend, I'd probably look at wrapping rustic in a small gRPC wrapper that allows it to be used as a daemon. 

_The challenging bit I see as a potential blocker_ (but want to solve!) is how to represent configuration -- I don't want to end up reproducing all of rustic's configs e.g. https://docs.rs/rustic_core/latest/rustic_core/struct.RepositoryBackends.html or https://docs.rs/rustic_core/latest/rustic_core/struct.BackupOptions.html as protobuffers . In Backrest today restic's configuration is transparent to Backrest, I get this for free by allowing users to specify arbitrary env vars and flags and linking the restic documentation. I'd like to be able to do something similar for configuring rustic_core objects when in a "rustic" mode and perhaps leveraging any config loading / parsing you've already built.

Summing up -- I think my question would be: would you consider supporting protobuf, json, or otherwise serializable formats of all of these configuration objects (with published schemas) so that they can be configured generically by wrapper scripts? I'm also happy to look at budgeting time to contribute here if we can align on an approach!

With something like that, I think rustic support actually looks reasonably straightforward :)

**Reply by garethgeorge** [reply]
Following up: I see some reference to config and perhaps serialization in rustic's code already e.g. 

 * https://github.com/rustic-rs/rustic_core/blob/main/crates/core/src/commands/config.rs#L106-L174 
 * https://github.com/rustic-rs/rustic_core/blob/main/examples/config/examples/config.rs 

Maybe that's most of the plumbing that I need / I'm being hampered by needing to familiarize myself a bit more with rust?

**Reply by aawsome** [reply]
@garethgeorge I think the best example for using rustic_core is something like https://github.com/rustic-rs/rustic_core/blob/main/examples/backup/examples/backup.rs.
Note that the config example shows how to change the in-repository config parameters like compression level and is not about config to open and access a repository.

What you basically need is [RepositoryOptions](https://docs.rs/rustic_core/latest/rustic_core/struct.RepositoryOptions.html) and [BackendOptions](https://docs.rs/rustic_backend/latest/rustic_backend/choose/struct.BackendOptions.html). Both already implement `serde::Serialize` and `serde::Deserialize`. They are used as TOML but other serialization formats should also work directly.

**Reply by garethgeorge** [reply]
Thanks for the very fast response especially after my late reply. Great -- let me give this a go tonight and see how far I can get.

Do any of those examples cover use of progress bars / log reporting?

**Reply by aawsome** [reply]
For progress bars, you simply have to implement the [ProgressBars](https://docs.rs/rustic_core/latest/rustic_core/trait.ProgressBars.html) trait. Note that the methods-to-implement return types which must implement the [Progress](https://docs.rs/rustic_core/latest/rustic_core/progress/trait.Progress.html) trait. An example how to implement progress bars using the `indicatif` crate to actually show the progress can be seen [here](https://github.com/rustic-rs/rustic/blob/main/src/config/progress_options.rs) - but you'll want to reimplement this such that it will send the information over gRPC.

For logs, rusitc_core currently uses the `log` crate. You might want to use [set_logger](https://docs.rs/log/latest/log/fn.set_logger.html) to implement handling of the logs.

**Reply by garethgeorge** [reply]
Hey, thanks for the details. I'll keep updating this thread as I get through my prototyping if it's not spammy :)

I had a chance to start prototyping with rustic and it looks very promising (even with my relatively limited rust experience -- I think the way to go here will just be keeping the operational model simple). rustic_core is nice to work with and gives me a phenomenal level of control compared to wrapping restic from the outside. I expect support will be experimental (and heavily advertised as such in the UI!) but seems worth while to keep rustic growing along side restic and to provide a very easy way to get started with either option -- at the end of the day I'm impressed with the potential rustic offers re: performance as an embeddable engine. I'm also highly appreciative that with rustic I should be able to offer some features that i can't with restic e.g. fast listings and file browsing via direct API access.

I'll circle back once I have more by way of prototyping the rust gRPC daemon embedding rustic_core and, if you're open to it, may be able to trickle back some feature requests once I'm able to play with it in Backrest.

**Reply by aawsome** [reply]
Great to hear and I'm very excited to see your progress - so, please keep us updated!
And, of course, if you have any question, please don't hesitate to ask!

Note that we just released rustic_core 0.3.1 which fixed a few smaller bugs and does export some necessary traits about Repository state which you may need if you want to write the type of a Repository, e.g. to write abstract methods.

As for file browsing, if you are talking about the contents of a file, have a look at [OpenFile::from_node](https://docs.rs/rustic_core/latest/rustic_core/vfs/struct.OpenFile.html#method.from_node) which gives you the `read_at` method to arbitrarily read parts of the file. You can have a look at https://github.com/rustic-rs/rustic/pull/1216 where I implemented a very basic version for the TUI included in rustic.

BTW: Any contribution including feature requests is always welcome!

**Reply by aawsome** [reply]
Ah, one point as you are talking about a daemon: Long-time existence of the `Repository` object is not yet fully supported. I mean, of course you can keep it as long as you wish and operations should still work. But if the real repository changes, there is currently no way to update `Repository`. Some operations are directed to the backend and will give you always up-to-date results. But all internal data (especially the index) has currently just the initial state and there is yet no way to update it - except of re-initializing a `Repository` object with all it means.
Note that this also applies to commands which are execute on that `Repository`. `Repository::backup` will for instance not update the index of `Repository`! You have to self-update `Repository`, e.g. by just re-initializing a new one.

So you should have this in mind. But I also think you already thought about that issue as you have the same with all data you get from a restic CLI call and store internally. There must be some way to handle `backup` runs which happened in-between - or even `forget` or `prune` runs...

Still, this is something I already had in my back-mind to implement as it offers quite promising options: The `Repository` could just detect the data which changed in the real repository and only re-read the needed stuff which would be pretty fast...

**Reply by garethgeorge** [reply]
I wasn't sure if rustic would handle this gracefully, but will aim to be pretty conservative with the deamon design :) I was planning on circling back for some "official design review" but I'll also drop a bit of my brain storming notes here --

The design direction I'm currently leaning towards is a gRPC deamon with a stateless interface. Every RPC will take the full configuration it needs to execute in its arguments which makes Backrest the source of truth. This means I'm thinking RPCs will look something like

```
message RepoContext {
    string repo_opts_json = 1; // serde serialized form of 
    string backend_opts_json = 1; // serde serialized form of https://docs.rs/rustic_backend/latest/rustic_backend/choose/struct.BackendOptions.html 
}

message BackupRequest {
    RepoContext repo_context = 1;
    string parent_snapshot_id = 2;
    string backup_opts_overrides_json = 3;
}

message BackupResponseStreamItem {
    // TBD but this will be a gRPC stream item
} 
```

```
service RusticDaemon {
    rpc ValidateRepoContext(RepoContext) ... // validate a command, needed to delegate asserting that objects are deserializable and wellformed to rust / any validation that rustic will implement.
    rpc Backup(BackupRequest) returns (stream BackupResponse) {} 
}
```

My thinking is possibly (at least for initial work) to spawn a daemon per backrest task that is responsible for any RPCs executed during that flow (or perhaps even per operation). In this way, I can have rustic just log to stdout and capture it the same way I do for restic today & it at least reduces the risk of the class of lifecycle bugs you talk about. Since rustic_core is primarily used in a CLI it's probably best for me not to break any "it's typically short lived" or "scoped to a few operations" assumptions.

I suppose one interesting question to ask is -- as I'm thinking of heading off in this daemon direction, does this make sense to you? The alternative would be to upstream config-less versions of the rustic commands (i've seen references to a config format, but it looks like everything is also supported via env vars?) and possibly json progress support to rustic itself and wrap via the CLI the same way I do restic. There are some advantages with this approach e.g. any UX improvements or new flags you add users will get for free.

If the process won't be long lived it's possible what I'm thinking re: daemon makes less sense than simply upstreaming some scriptability features to the CLI

**Reply by garethgeorge** [reply]
Followinp up re: wrapping at the CLI level, I think the only capabilities that are strictly required for an MVP are

 * ability to parse the ID of the last created snapshot out of the backup command (and ideally parsable progress though I could simply show the TUI text in the UI).
 * parseable output from snapshot listings 

As a followup I'd also want

 * parseable output from forget 
 * parseable output from operations for browsing snapshots

**Reply by aawsome** [reply]
For your BackupRequest, [BackupOptions](https://docs.rs/rustic_core/latest/rustic_core/struct.BackupOptions.html) is also seralizable and includes parent related options.

> I suppose one interesting question to ask is -- as I'm thinking of heading off in this daemon direction, does this make sense to you?

This actually makes perfectly sense to me. You just have to make sure that sometimes you have to re-open the `Repository` instead of using the existing one. For each backup request, you should do
```Rust
let repo = Repository::new(&repo_opts, &backends)?
        .open()?
        .to_indexed_ids()?;
```
before calling `repo.backup()`, for example. And you can directly throw away that `repo` after that call.
This mimics the CLI-oriented design rustic_core still has. But it will be still more efficient and flexible than a CLI-call would be. Also the libraries would all be compiled into your daemon binary with no need of an additional binary. Plus, we may soon see some `repo.update()` in rustic_core which will be much more efficient and will allows you to adapt to with only small code changes.

Moreover, your daemon could allow for navigating around dirs in a snapshot. Here, the `repo` can be kept and re-used for each user action.

**Reply by aawsome** [reply]
> Followinp up re: wrapping at the CLI level, I think the only capabilities that are strictly required for an MVP are
> 
>     * ability to parse the ID of the last created snapshot out of the backup command (and ideally parsable progress though I could simply show the TUI text in the UI).
> 
>     * parseable output from snapshot listings
> 
> 
> As a followup I'd also want
> 
>     * parseable output from forget
> 
>     * parseable output from operations for browsing snapshots

For the rustic CLI, you don't have parseable progress and to parse `rustic ls` you have to parse the text output. All other requirements are fulfilled. (And you get all when using rustic_core)

**Reply by aawsome** [reply]
About efficient updating of snapshots, there is now https://github.com/rustic-rs/rustic_core/pull/260.
For efficient index updating, see https://github.com/rustic-rs/rustic_core/issues/261

**Reply by garethgeorge** [reply]
Thanks for those updates! Spent a bit more time experimenting this weekend

Now that I've had some time to prototype out the gRPC approach (which is fairly minimal so far but gives me a feel for what I'd need to maintain in terms of a daemon process) https://github.com/garethgeorge/backrest/pull/462 I'm becoming torn on the right way to proceed, I see some cons with reinventing the wheel with the gRPC daemon 

  * I think a lot of users like Backrest because it's such a thin wrapper e.g. if they're familiar with restic, that's also the behavior they get from Backrest. Not sure whether it makes sense for Backrest to be in the business of building an opinionated engine that diverges from the base project.
  * Build system complexity: I'll need to manage and ship releases of the daemon from some project. Right now I'm building it in a mono-repo but it probably needs to be a separate project with it's own pipelines and release schedule.
  * Configuration divergence -- rustic supports a number of flags and env vars in its CLI which are well designed and closely mirror the options restic exposes, I want to keep that familiarity / pass it on to users and am realizing anything I do here will duplicate work you're already doing on the CLI.
  * There isn't a direct pathway to support the "Run Command" feature which lets users craft their own CLI invocations of "restic" today or "rustic" in rustic mode.

The obvious pro of the daemon / building on rustic_core is the level of control but it's arguable that Backrest is trying to be closer to a scripting layer than an actual backup engine!

I pivoted a bit to spend some time playing around with the rustic CLI itself and I'd somehow entirely missed the JSON support in my first look at the project. If I switch rustic to log to a file (logs sometimes interrupt JSON output) it looks likely to be straightforward to parse and the output you generate is already relatively possible to transform into the structs I use for restic.

Wrinkles I see with automating by wrapping the CLI are pretty slight:

 * Lack of JSON progress is a bit tricky but I'm imagining JSON progress bars are something I could PR for a future release I suspect
 * Logs occasionally print while producing JSON output -- will likely be addressed by just logging to a file instead of stdout.

Otherwise CLI behavior is very close to restic today and I can likely support both under a single interface. Aiming to find some time this week to prototype _this_ as if it's as easy as it's looking I might be able to get this up and running in a few days.

So current status is -- likely I'll continue a bit further with the gRPC approach but I'm also going to look at just wrapping the CLI and see how stable I can make that. If it is stable this is probably my preference barring upstreaming a gRPC daemon to rustic. WDYT?

**Reply by aawsome** [reply]
Thanks for keeping us up-to-date @garethgeorge 
I can understand your considerations. And of course you can use the rustic CLI itself and it will be able to replace restic in most circumstances.

About JSON progress: This can IMO be implemented pretty easily by using another `ProgressBars` implementation when `--json` is specified. If you want to work on this, I can support you.

About logs: you can log to a file or use the `--log-level` option: You may want to set it to `error` or `warn`. But also note that logs go to stderr (like progress information) whereas json output goes to stdout. If you take this into account you may just use rustic as-is - maybe just directly show stderr to the user and read/interpret stdout json.

**Comment by simonsan** [comment]
One thing hasn't been mentioned here, there are two libraries you can use. One is `rustic_core` which is the low level library. The other one is `rustic-rs` which is the cli library - you find it in the main repo under `/src`. Main.rs is the binary (a very thin wrapper around the lib) and everything else is the library. We haven't used it for that (because we really just call into the apps entrypoint) but we probably can make it possible to call the high level commands in code. Maybe you can look into that approach and see if you need any changes there to make your use case possible. Most Runnable commands can be also called from outside of the trait implementation via run_inner/inner_run. I would recommend using your own branch of that library and PR the changes you needed to make to make it work, a lot of items you will need are currently not visible/pub. You can change that and we discuss about it and what other changes are needed. I might invest a bit of time after the weekend to decouple some logic there from the commands. So they become easier to use without abscissa and can be easier tested as well.

**Comment by f18m** [comment]
Hey @garethgeorge ,
I wonder if you have news about supporting Rustic from Backrest... I would be very interested in that...

**Reply by LukasdeBoer** [reply]
This would be a great addition indeed. As far as I know there are no backup tools with a web GUI that support S3 deep glacier natively and this would be a game changer for both tools!

**Reply by xxxliqu1dxxx** [reply]
@garethgeorge it would be indeed great if you could share an update - lots of us would like to play with backrest with rustic in the backend... perhaps it should be called backrust? :)

**Reply by maelstrm-code** [reply]
@garethgeorge this conversation looks very interesting, voting +1 that this would be an appreciated improvement. Locks are currently a pain in restic and one of the advantages of rustic appears to be no locks!
